### PR TITLE
Automate creating html from tsv file

### PR DESCRIPTION
**Pull request by janwadolowski**
#### When merged, this PR will:
- Add parser for TSV files generating HTML strings from them

### CODE REVIEW

File: src/sutta_publisher/ingester/base.py [hunk 1]
```diff
@@ -13,7 +14,10 @@ class Ingester:
 
     @classmethod
     def get_result(cls) -> IngestResult:
-        log.info("** Running conversion for publication: %s", cls.config.publication_number)
-        result = IngestResult(content=f"<html><body>{cls.config.publication_number=}</body></html>")
-        log.info("** Finished conversion for publication: %s", cls.config.publication_number)
+        # log.info("** Running conversion for publication: %s", cls.config.publication_number)
```
Comment (alple) [inline src/sutta_publisher/ingester/base.py:17]: remove unused code

File: src/sutta_publisher/ingester/base.py [hunk 2]
```diff
@@ -13,7 +14,10 @@ class Ingester:
...
+        parser = TsvParser("dn.tsv")
+        # result = IngestResult(content=f"<html><body>{cls.config.publication_number=}</body></html>")
+        result = IngestResult(content=f"<html><body>{parser.parse_input()}</body></html>")
+        # log.info("** Finished conversion for publication: %s", cls.config.publication_number)
+        print(parser.parse_input())
```
Comment (alple) [inline src/sutta_publisher/ingester/base.py:22]: Print is unnecessary, printing or dumping to file should be done in publisher

File: src/sutta_publisher/ingester/helper_functions.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+
+def _fetch_possible_refs() -> list[str]:
+    import json
+    from urllib.request import urlopen
+
+    all_references_list_url = (
+        "https://raw.githubusercontent.com/suttacentral/sc-data/master/misc/pali_reference_edition.json"
+    )
+    response = urlopen(all_references_list_url)
```
Comment (alple) [inline src/sutta_publisher/ingester/helper_functions.py:12]: why not using `requests`?

File: src/sutta_publisher/ingester/tsv_parser.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        self.spreadsheet = pd.read_csv(tsv, sep="\t", header=0)
+        # columns with translations have translator's name in it (eg. translation-en-sujato) so we need regex to find English translation
+        en = _catch_translation_en_column(self.spreadsheet.columns)
+        self.spreadsheet = self.spreadsheet[[en, "html", "reference", "segment_id"]]
+        self.spreadsheet.rename(columns={en: "english"}, inplace=True)
+        self.spreadsheet.fillna("")  # replace empty cells with empty strings, so we can assume all input has str
+
+    def parse_input(self) -> str:
+        """Parse a tsv spreadsheet content and construct HTML string"""
+        # TODO: replace with Jinja2 template
```
Comment (alple) [inline src/sutta_publisher/ingester/tsv_parser.py:27]: If you are adding todo, also add an issue for that and point to that issue in this line

File: src/sutta_publisher/ingester/tsv_parser.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        output = ""
+        # TODO: look for more efficient method for dataframes than looping
+        for _, row in self.spreadsheet.iterrows():
+            html: list[str] = _split_html_on_bracket(row["html"])
+            split_ref: tuple[str, str] = _split_ref_and_number(row["reference"])
+            reference: str = _reference_to_html(split_ref) if split_ref else ""
+            segment_id = _segment_id_to_html(row["segment_id"])
+            # e.g.      <p><a class='sc-main' id='dn1:1.10.16'>DN 1:1.10.16</a>‘He refrains from running errands...<a id='dn1:1.10.17'></a></p>
+            output += f"{html[0]}{reference}{row['english']}{segment_id}{html[1]}"
+        return output
```
Comment (alple) [inline src/sutta_publisher/ingester/tsv_parser.py:38]: Lets make `output` it to the list, should be less churning for python: 

```suggestion
            output.append(f"{html[0]}{reference}{row['english']}{segment_id}{html[1]}")
        return "".join(output)
```

File: tests/unit/test_tsv_parser.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
+from io import StringIO
+
+from sutta_publisher.ingester.tsv_parser import TsvParser
+
+
+def test_should_generate_valid_html_string():
+    input_tsv = StringIO(
```
Comment (alple) [inline tests/unit/test_tsv_parser.py:7]: Let's put them into `csv` and `html` files

File: tests/unit/test_helper_functions.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+        {"edition_set": "km", "includes": "km", "name": "Phratraipiṭakapāḷi (Cambodia), 1958–1969"},
+        {"edition_set": "lv", "includes": "lv", "name": "Lāvaraṭṭhassa Tipiṭaka, 1957"},
+        {"edition_set": "ndp", "includes": "ndp", "name": "Nālandā Devanāgarī Pāḷi Series Tipiṭaka, 1957–1962"},
+        {"edition_set": "cck", "includes": "cck", "name": "Chulachomklao Pāḷi Tipiṭaka, 1893"},
+        {"edition_set": "sya", "includes": ["sya1ed", "sya2ed", "sya-all"], "name": "Other Thai editions"},
+        {"edition_set": "vri", "includes": "vri", "name": "Vipassanā Research Institute Tipiṭaka, 2537–2542"},
+        {"edition_set": "maku", "includes": "maku", "short_name": "Mahāmakut (Milindapañha), 1923"},
+    ]
+    # patch internal func calls, so we don't need to actually download json from the Internet
+    mocker.patch(target="urllib.request.urlopen")
```
Comment (alple) [inline tests/unit/test_helper_functions.py:84]: Lets use `pytest-vcr`

**Comment by janwadolowski** [comment]
Many changes have been introduced thanks to Alex' comments and pair coding, please re-review if anything has been missed.

I'm changing the PR status from draft to normal as it's a working solution (althouth still with its flows and compromises)

**Comment by janwadolowski** [comment]
For now it seems that idea with TsvParser is dead since we will get the data from API in a form of json, not tsv. Useful stuff will be cheery picked from this branch into [22-create-json-parser](https://github.com/suttacentral/publications/tree/22-create-json-parser)

**Pull request closed without merging**